desc/geometry/surface.py

```diff
@@ -890,3 +893,315 @@ def set_coeffs(self, l, m=0, R=None, Z=None):
+                modes_R = np.array([[0, 0], [1, 1]])
+            if Z_lmn is None:
+                Z_lmn = np.array([0, -1])
+                modes_Z = np.array([[0, 0], [1, -1]])
+            if L_lmn is None:
+                L_lmn = np.array([0, -1])
+                modes_L = np.array([[0, 0], [1, -1]])
+            if modes_L is None:
+                modes_L = modes_R
+            if modes_Z is None:
+                modes_Z = modes_R
+
+            L_lmn, modes_R, modes_Z, modes_L = map(
+                np.asarray, (L_lmn, modes_R[:, :2], modes_Z[:, :2], modes_L[:, :2])
+            )
+
+            assert (
+                L_lmn.size == modes_L.shape[0]
+            ), "L_lmn size and modes_L.shape[0] must be the same size!"
+
+            assert issubclass(modes_L.dtype.type, np.integer)
+
+            super().__init__(
+                R_lmn=R_lmn,
+                Z_lmn=Z_lmn,
+                modes_R=modes_R,
+                modes_Z=modes_Z,
+                spectral_indexing=spectral_indexing,
+                sym=sym,
+                zeta=zeta,
+                name=name,
+                check_orientation=check_orientation,
+            )
+
+            LR = np.max(abs(modes_R[:, 0]))
+            MR = np.max(abs(modes_R[:, 1]))
+            LZ = np.max(abs(modes_Z[:, 0]))
+            MZ = np.max(abs(modes_Z[:, 1]))
+            LL = np.max(abs(modes_L[:, 0]))
+            ML = np.max(abs(modes_L[:, 1]))
+
+            self._L = max(self.L, LL)
+            self._M = max(self.M, ML)
+            self._N = 0
+
+            self._R_basis = ZernikePolynomial(
+                L=max(LR, MR),
+                M=MR,
+                spectral_indexing=spectral_indexing,
+                sym="cos" if sym else False,
+            )
+            self._Z_basis = ZernikePolynomial(
+                L=max(LZ, MZ),
+                M=MZ,
+                spectral_indexing=spectral_indexing,
+                sym="sin" if sym else False,
+            )
+            self._L_basis = ZernikePolynomial(
+                L=max(LL, ML),
+                M=ML,
+                spectral_indexing=spectral_indexing,
+                sym="sin" if sym else False,
+            )
+
+            self._R_lmn = copy_coeffs(R_lmn, modes_R, self.R_basis.modes[:, :2])
+            self._Z_lmn = copy_coeffs(Z_lmn, modes_Z, self.Z_basis.modes[:, :2])
+            self._L_lmn = copy_coeffs(L_lmn, modes_L, self.L_basis.modes[:, :2])
+            self.name = name
+
+    @property
+    def L_basis(self):
+        """ZernikePolynomial: Spectral basis for Lambda."""
+        return self._L_basis
+
+    @L_basis.setter
+    def L_basis(self, L_basis):
+        self._L_basis = L_basis
+
+    @optimizable_parameter
+    @property
+    def L_lmn(self):
+        """ndarray: Spectral coefficients for Lambda."""
+        return self._L_lmn
+
+    @L_lmn.setter
+    def L_lmn(self, L_lmn):
+        self._L_lmn = L_lmn
+
+    @property
+    def isgiven(self):
```

Comment:
this feels really weird to have this be an attribute of the class? Like, the section object shouldnt have to know how its being used

Reply:
I got rid of isgiven completely